include/rocksdb/utilities/secondary_index.h

```diff
@@ -0,0 +1,100 @@
+// per-column basis. The secondary index then maintains a mapping from (column)
+// value to the list of primary keys that have the corresponding value (in the
+// given column).
+//
+// The primary and secondary key-value pairs can be stored in either the same
+// column family or different ones. It is the application's responsibility to
+// avoid conflicts and ambiguities (for example, by using prefixes to create
+// separate key spaces).
+//
+// In the general case where there are concurrent writers, maintaining a
+// secondary index requires transactional semantics and concurrency control.
+// Because of this, secondary indices are only supported via the transaction
+// layer. With secondary indices, whenever a (primary) key-value is inserted,
+// updated, or deleted via a transaction (regardless of whether it is an
+// explicit or implicit one), RocksDB will invoke any applicable SecondaryIndex
+// objects based on primary column family and column name, and it will
+// automatically add or remove any secondary index entries as needed (using
+// the same transaction).
+//
+// Note: the methods of SecondaryIndex implementations are expected to be
+// thread-safe with the exception of Set{Primary,Secondary}ColumnFamily (which
+// are not expected to be called after initialization).
+class SecondaryIndex {
+ public:
+  virtual ~SecondaryIndex() = default;
+
+  virtual void SetPrimaryColumnFamily(ColumnFamilyHandle* column_family) = 0;
+  virtual void SetSecondaryColumnFamily(ColumnFamilyHandle* column_family) = 0;
+
+  virtual ColumnFamilyHandle* GetPrimaryColumnFamily() const = 0;
+  virtual ColumnFamilyHandle* GetSecondaryColumnFamily() const = 0;
+
+  // The name of the primary column to index. Plain key-values can be indexed by
+  // specifying kDefaultWideColumnName.
+  virtual Slice GetPrimaryColumnName() const = 0;
+
+  // Optionally update the primary column value during an insert or update of a
+  // primary key-value. Called by the transaction layer before the primary
+  // key-value write is added to the transaction. Returning a non-OK status
+  // rolls back all operations in the transaction related to this primary
+  // key-value.
+  virtual Status UpdatePrimaryColumnValue(
+      const Slice& primary_key, const Slice& primary_column_value,
+      std::optional<std::variant<Slice, std::string>>* updated_column_value)
+      const = 0;
+
+  // Get the secondary key prefix for a given primary key-value. This method is
+  // called by the transaction layer when adding or removing secondary index
+  // entries (which have the form <secondary_key_prefix><primary_key> ->
+  // <secondary_value>) and should be deterministic. The output parameter
+  // secondary_key_prefix is expected to be based on primary_column_value,
+  // potentially with some additional metadata to prevent ambiguities (e.g.
+  // index id or length indicator). Returning a non-OK status rolls back all
```

Comment:
Thanks for the review and the questions @jowlyzhang !

> I wonder if this SecondaryIndex interface is for creating just one index. If yes, would it make sense to have a `GetIndexId` API, and concatenate it for all the <secondary_key_prefix>.

Yes, it corresponds to a single index. Regarding index ids, if an index has its very own column family, having the id in each key is not really necessary, so I thought I'd keep this flexible and defer to the implementation as to whether an id is needed.

> Is this interface completely in memory, or its own metadata needs to be persisted somewhere and reused across DB sessions. In that case, would it also make sense to have something like `GetIndexName` API.

> I wonder whether users will need RocksDB support to answer questions like, given a primary ColumnFamilyHandle*, which column family is holding its secondary index "a" etc. Maybe that's not something this interface itself need to directly support and will be in a different layer.

Yeah, on the long run I think we would probably want to keep track of indices in the MANIFEST and automatically instantiate the appropriate `SecondaryIndex` objects. This would also involve storing the primary and secondary column families of indices like you suggested. However, in the initial version, I think we can leave it up to the application to provide the `SecondaryIndex` instances (as part of `TransactionDBOptions`, see D66672931), the same way they e.g. provide the table factory.